### PR TITLE
add AsyncGenerator

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Looks like this was inadvertently omitted from 3.6 as released.

Should I submit a separate patch to CPython directly to add documentation?

### CODE REVIEW

**Comment by gvanrossum** [comment]
Thanks! I will merge this after 3.5.3 has been released.

**Comment by gvanrossum** [comment]
And yes, please send a separate doc patch.

**Comment by JelleZijlstra** [comment]
http://bugs.python.org/issue29198

File: src/typing.py [hunk 1]
```diff
@@ -1859,6 +1860,21 @@ def __new__(cls, *args, **kwds):
...
         return _generic_new(_G_base, cls, *args, **kwds)
 
+if hasattr(collections_abc, 'AsyncGenerator'):
+    _AG_base = collections_abc.AsyncGenerator
+
+    class AsyncGenerator(AsyncIterator[T_co], Generic[T_co, T_contra],
+                         extra=_AG_base):
+        __slots__ = ()
+
+        def __new__(cls, *args, **kwds):
```
Comment (ilevkivskyi) [inline src/typing.py]: Since you always inherit from ``collections.abc`` there is no need to override ``__new__``. It will be anyway not possible to instantiate ``typing.AsyncGenerator`` because ``collections.abc.AsyncGenerator`` has abstract methods.

Comment (JelleZijlstra) [inline src/typing.py]: This was mostly copied from the `Generator` implementation just above. It is possible to instantiate a subclass of `AsyncGenerator`, which would use this `__new__`.

Comment (ilevkivskyi) [inline src/typing.py]: I don't think I understand your point. The only way to instantiate a subclass of an ABC is to implement all abstract methods. Moreover, there is no way a subclass could "use" this ``__new__``, it only exists everywhere in ``typing`` to prohibit direct instantiation of generic version of concrete collections and similar things.

``__new__`` for ``Generator`` is necessary, since it could inherit from ``types.GeneratorType``.

File: src/test_typing.py [hunk 2]
```diff
@@ -1320,6 +1320,7 @@ def blah():
 
 
 ASYNCIO = sys.version_info[:2] >= (3, 5)
+ASYNC_GENERATOR = sys.version_info[:2] >= (3, 6)
```
Comment (gvanrossum) [inline src/test_typing.py]: Why is this line needed? The variable is not used AFAICT.

Comment (JelleZijlstra) [inline src/test_typing.py]: Yes, looks like I didn't end up needing it. I'll remove it.

**Comment by gvanrossum** [comment]
@ilevkivskyi Are you happy with this code now? Then I can merge it. (Honestly I want to give you commit privs here if you don't already have them.)

**Comment by ilevkivskyi** [comment]
> Are you happy with this code now?

There are two very minor points: the variable ``_AG_base`` is used only once, and the last test ``self.assertNotIsInstance(type(g), G)`` is not very useful, I would add ``self.assertNotIsInstance(g, G)``. Otherwise I am happy.

> Honestly I want to give you commit privs here if you don't already have them.

I am glad that you trust me. I don't have the privileges. I think it indeed may simplify things.

**Comment by gvanrossum** [comment]
I invited you. Go ahead and commit this PR when you are happy with it.

**Comment by ilevkivskyi** [comment]
Probably the best way is to merge this now and I will add those minor changes myself, so that #355 will be "unblocked" very soon.
